url147ce07ebe_JoshuaKissoon_Kademlia_tgz-pJ8-kademlia_simulations_RoutingTableStateTestingJ8/src/kademlia/JKademliaNode.java

```diff
@@ -0,0 +1,429 @@
+                    JKademliaNode.this.refresh();
+                }
+                catch (IOException e)
+                {
+                    System.err.println("KademliaNode: Refresh Operation Failed; Message: " + e.getMessage());
+                }
+            }
+        };
+        refreshOperationTimer.schedule(refreshOperationTTask, this.config.restoreInterval(), this.config.restoreInterval());
+    }
+
+    @Override
+    public final void stopRefreshOperation()
+    {
+        /* Close off the timer tasks */
+        this.refreshOperationTTask.cancel();
+        this.refreshOperationTimer.cancel();
+        this.refreshOperationTimer.purge();
+    }
+
+    public JKademliaNode(String ownerId, Node node, int udpPort, KademliaRoutingTable routingTable, KadConfiguration config) throws IOException
+    {
+        this(
+                ownerId,
+                node,
+                udpPort,
+                new DHT(ownerId, config),
+                routingTable,
+                config
+        );
+    }
+
+    public JKademliaNode(String ownerId, Node node, int udpPort, KadConfiguration config) throws IOException
+    {
+        this(
+                ownerId,
+                node,
+                udpPort,
+                new JKademliaRoutingTable(node, config),
+                config
+        );
+    }
+
+    public JKademliaNode(String ownerId, KademliaId defaultId, int udpPort) throws IOException
+    {
+        this(
+                ownerId,
+                new Node(defaultId, InetAddress.getLocalHost(), udpPort),
+                udpPort,
+                new DefaultConfiguration()
+        );
+    }
+
+    /**
+     * Load Stored state using default configuration
+     *
+     * @param ownerId The ID of the owner for the stored state
+     *
+     * @return A Kademlia instance loaded from a stored state in a file
+     *
+     * @throws java.io.FileNotFoundException
+     * @throws java.lang.ClassNotFoundException
+     */
+    public static JKademliaNode loadFromFile(String ownerId) throws FileNotFoundException, IOException, ClassNotFoundException
+    {
+        return JKademliaNode.loadFromFile(ownerId, new DefaultConfiguration());
+    }
+
+    /**
+     * Load Stored state
+     *
+     * @param ownerId The ID of the owner for the stored state
+     * @param iconfig Configuration information to work with
+     *
+     * @return A Kademlia instance loaded from a stored state in a file
+     *
+     * @throws java.io.FileNotFoundException
+     * @throws java.lang.ClassNotFoundException
+     */
+    public static JKademliaNode loadFromFile(String ownerId, KadConfiguration iconfig) throws FileNotFoundException, IOException, ClassNotFoundException
+    {
+        DataInputStream din;
+
+        /**
+         * @section Read Basic Kad data
+         */
+        din = new DataInputStream(new FileInputStream(getStateStorageFolderName(ownerId, iconfig) + File.separator + "kad.kns"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *din*. Currently, there are execution paths that do not contain closure statements, for example, when JsonSerializer.read throws an exception. Ensure *din* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).